File: go/hash/base32_test.go
// Copyright 2016 Attic Labs, Inc. All rights reserved.
// Licensed under the Apache License, version 2.0:
// http://www.apache.org/licenses/LICENSE-2.0

package hash

import (
	"testing"

	"github.com/attic-labs/testify/assert"
)

func TestBase32Encode(t *testing.T) {
	assert := assert.New(t)

	d := make([]byte, 20, 20)
	assert.Equal("00000000000000000000000000000000", encode(d))
	d[19] = 1
	assert.Equal("00000000000000000000000000000001", encode(d))
	d[19] = 10
	assert.Equal("0000000000000000000000000000000a", encode(d))
	d[19] = 20
	assert.Equal("0000000000000000000000000000000k", encode(d))
	d[19] = 31
	assert.Equal("0000000000000000000000000000000v", encode(d))
	d[19] = 32
	assert.Equal("00000000000000000000000000000010", encode(d))
	d[19] = 63
	assert.Equal("0000000000000000000000000000001v", encode(d))
	d[19] = 64
	assert.Equal("00000000000000000000000000000020", encode(d))

	// Largest!
	for i := 0; i < 20; i++ {
		d[i] = 0xff
	}
	assert.Equal("vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv", encode(d))
}

func TestBase32Decode(t *testing.T) {
	assert := assert.New(t)

	d := make([]byte, 20, 20)
	assert.Equal(d, decode("00000000000000000000000000000000"))

	d[19] = 1
	assert.Equal(d, decode("00000000000000000000000000000001"))
	d[19] = 10
	assert.Equal(d, decode("0000000000000000000000000000000a"))
	d[19] = 20
	assert.Equal(d, decode("0000000000000000000000000000000k"))
	d[19] = 31
	assert.Equal(d, decode("0000000000000000000000000000000v"))
	d[19] = 32
	assert.Equal(d, decode("00000000000000000000000000000010"))
	d[19] = 63
	assert.Equal(d, decode("0000000000000000000000000000001v"))
	d[19] = 64
	assert.Equal(d, decode("00000000000000000000000000000020"))

	// Largest!
	for i := 0; i < 20; i++ {
		d[i] = 0xff
	}
	assert.Equal(d, decode("vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv"))
}

File: js/noms/src/base32-test.js
// @flow

// Copyright 2016 Attic Labs, Inc. All rights reserved.
// Licensed under the Apache License, version 2.0:
// http://www.apache.org/licenses/LICENSE-2.0

import {encode, decode} from './base32.js';
import {assert} from 'chai';
import {alloc} from './bytes.js';
import {suite, test} from 'mocha';

suite('base32', () => {
  test('encode', () => {
    const a = alloc(20);
    assert.equal(encode(a).length, 32);
    assert.equal(encode(a), '00000000000000000000000000000000');
    a[a.length - 1] = 1;
    assert.equal(encode(a), '00000000000000000000000000000001');
    a[a.length - 1] = 10;
    assert.equal(encode(a), '0000000000000000000000000000000a');
    a[a.length - 1] = 20;
    assert.equal(encode(a), '0000000000000000000000000000000k');
    a[a.length - 1] = 31;
    assert.equal(encode(a), '0000000000000000000000000000000v');
    a[a.length - 1] = 32;
    assert.equal(encode(a), '00000000000000000000000000000010');
    a[a.length - 1] = 63;
    assert.equal(encode(a), '0000000000000000000000000000001v');
    a[a.length - 1] = 64;
    assert.equal(encode(a), '00000000000000000000000000000020');

    // Largest!
    for (let i = 0; i < a.length; i++) {
      a[i] = 0xff;
    }
    assert.equal(encode(a), 'vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv');
  });

  test('decode', () => {
    assert.equal(decode('00000000000000000000000000000000').length, 20);

    const a = alloc(20);
    assert.deepEqual(decode('00000000000000000000000000000000'), a);

    a[a.length - 1] = 1;
    assert.deepEqual(decode('00000000000000000000000000000001'), a);
    a[a.length - 1] = 10;
    assert.deepEqual(decode('0000000000000000000000000000000a'), a);
    a[a.length - 1] = 20;
    assert.deepEqual(decode('0000000000000000000000000000000k'), a);
    a[a.length - 1] = 31;
    assert.deepEqual(decode('0000000000000000000000000000000v'), a);
    a[a.length - 1] = 32;
    assert.deepEqual(decode('00000000000000000000000000000010'), a);
    a[a.length - 1] = 63;
    assert.deepEqual(decode('0000000000000000000000000000001v'), a);
    a[a.length - 1] = 64;
    assert.deepEqual(decode('00000000000000000000000000000020'), a);

    // Largest!
    for (let i = 0; i < a.length; i++) {
      a[i] = 0xff;
    }
    assert.deepEqual(decode('vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv'), a);
  });
});
